adapters/lora-vertx/src/main/java/org/eclipse/hono/adapter/lora/LoraProtocolAdapter.java

```diff
@@ -0,0 +1,636 @@
+                TracingHelper.logError(getCurrentSpan(ctx), "Got message without deviceId");
+            }
+            if (payloadBase64 == null) {
+                TracingHelper.logError(getCurrentSpan(ctx), "Got message without valid payload");
+            }
+            handle400(ctx, JSON_MISSING_REQUIRED_FIELDS);
+        }
+    }
+
+    private JsonObject createDownstreamJson(final String base64Payload) {
+        return new JsonObject().put("payload", base64Payload);
+    }
+
+    private void handle202(final RoutingContext ctx) {
+        Tags.HTTP_STATUS.set(getCurrentSpan(ctx), 202);
+        ctx.response().setStatusCode(202);
+        ctx.response().end();
+    }
+
+    private void handle401(final RoutingContext ctx) {
+        Tags.HTTP_STATUS.set(getCurrentSpan(ctx), 401);
+        HttpUtils.unauthorized(ctx, "Basic realm=\"" + getConfig().getRealm() + "\"");
+    }
+
+    private void handle400(final RoutingContext ctx, final String msg) {
+        Tags.HTTP_STATUS.set(getCurrentSpan(ctx), 400);
+        HttpUtils.badRequest(ctx, msg);
+    }
+
+    private void handle200(final RoutingContext ctx) {
+        Tags.HTTP_STATUS.set(getCurrentSpan(ctx), 200);
+        ctx.response().setStatusCode(200);
+        ctx.response().end();
+    }
+
+    private void scheduleStartLoraCommandConsumers() {
+        final List<String> commandEnabledTenants = loraCommandProperties.getCommandEnabledTenants();
+        LOG.info("Starting Lora command consumers for tenants [{}] ...", commandEnabledTenants);
+        for (final String tenantId : commandEnabledTenants) {
+            scheduleStartLoraCommandConsumer(tenantId);
+        }
+    }
+
+    private void scheduleStartLoraCommandConsumer(final String tenantId) {
+        LOG.info("Starting Lora command consumer for tenant '{}' ...", tenantId);
+        startLoraCommandConsumer(tenantId).recover(x -> {
+            LOG.error("Error starting initial Lora command consumer for tenant [{}], retry in {} ms", tenantId,
+                    LORA_COMMAND_CONSUMER_RETRY_INTERVAL, x);
+            vertx.setTimer(LORA_COMMAND_CONSUMER_RETRY_INTERVAL, y -> scheduleStartLoraCommandConsumer(tenantId));
+            return Future.succeededFuture();
+        });
+    }
+
+    private Future<MessageConsumer> startLoraCommandConsumer(final String tenantId) {
+        return getCommandConnection().createCommandConsumer(tenantId, LORA_COMMAND_CONSUMER_DEVICE_ID,
+                receivedCommandContext -> commandConsumer(tenantId, receivedCommandContext), close -> {
+                    LOG.info("Lora command consumer closing initiated");
+                    closeCommandConsumer(tenantId, LORA_COMMAND_CONSUMER_DEVICE_ID);
+                }, LORA_COMMAND_CONSUMER_RETRY_INTERVAL);
+    }
+
+    @Override
+    protected void onStartupSuccess() {
+        scheduleStartLoraCommandConsumers();
+    }
+
+    private void commandConsumer(final String tenantId, final CommandContext receivedCommandContext) {
```

Comment:
you seem to already have the command direction implemented as well, is that right?